From 67cbcb0a70f99ec3d085ddfd6b3732ec4def4ee7 Mon Sep 17 00:00:00 2001
From: jzhuan5 <jin.can.zhuang@intel.com>
Date: Wed, 6 Nov 2013 11:22:21 +0800
Subject: [PATCH 195/429] dwc3-device: Fix maxburst issue

endpoint.maxburst may be 0 if a function doesn't call
config_ep_by_speed() to update it from the companion descriptor.
And endpoint.maxburst - 1 returns 11111b which wrongly sets bit
26 of endpoint parameter 0.
This sets a wrong endpoint state and will cause "Get Endpoint State"
command can't get the corret endpoint state and "Set Endpoint Config"
command can't retore the correct endpoint state during hibernation
resume flow.

Thus, when endpoint.maxburst is 0, we should set burst as 0 directly.

Signed-off-by: jzhuan5 <jin.can.zhuang@intel.com>
---
 drivers/usb/dwc3/gadget.c | 10 +++++++---
 1 file changed, 7 insertions(+), 3 deletions(-)

diff --git a/drivers/usb/dwc3/gadget.c b/drivers/usb/dwc3/gadget.c
index 893a95fbde9e..cade5a56709d 100644
--- a/drivers/usb/dwc3/gadget.c
+++ b/drivers/usb/dwc3/gadget.c
@@ -477,9 +477,13 @@ static int dwc3_gadget_set_ep_config(struct dwc3 *dwc, struct dwc3_ep *dep,
 		| DWC3_DEPCFG_MAX_PACKET_SIZE(maxp)
 		| cfg_action;
 
-	if (dep->ebc) {
-		if (dwc->gadget.speed == USB_SPEED_SUPER) {
-			u32 burst = 0;
+	/* Burst size is only needed in SuperSpeed mode */
+	if (dwc->gadget.speed == USB_SPEED_SUPER) {
+		/* In case a function forgets to set maxburst, maxburst may be
+		 * still 0, and we shouldn't minus 1 for it.
+		 */
+		u32 burst = dep->endpoint.maxburst ?
+				dep->endpoint.maxburst - 1 : 0;
 
 			params.param0 |= DWC3_DEPCFG_BURST_SIZE(burst);
 		}
-- 
2.37.3

